### PR TITLE
HugePages should require the user to request cpu or memory

### PR DESCRIPTION
**Pull request by tanshanshan**
**What this PR does / why we need it**:
 HugePages should require the user to request cpu or memory
ref #53672 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by tanshanshan** [comment]
cc @derekwaynecarr

File: pkg/api/validation/validation.go [hunk 1]
```diff
@@ -4016,6 +4029,16 @@ func ValidateResourceRequirements(requirements *api.ResourceRequirements, fldPat
...
 		}
+		if helper.IsHugePageResourceName(resourceName) {
+			reqContainsHugePages = true
+		}
+		if supportedQoSComputeResources.Has(string(resourceName)) {
+			reqContainsCpuOrMemory = true
+		}
+
+	}
+	if !limContainsCpuOrMemory && !reqContainsCpuOrMemory && (reqContainsHugePages || limContainsHugePages) {
```
Comment (tianshapjq) [inline pkg/api/validation/validation.go]: what if I don't specify the cpu, memory and hugepages in the yaml file, would this be error?

File: pkg/api/validation/validation_test.go [hunk 2]
```diff
@@ -2792,10 +2792,14 @@ func TestAlphaHugePagesIsolation(t *testing.T) {
...
-								api.ResourceName("hugepages-2Mi"): resource.MustParse("1Gi"),
+								api.ResourceName(api.ResourceCPU):    resource.MustParse("10"),
+								api.ResourceName(api.ResourceMemory): resource.MustParse("10G"),
+								api.ResourceName("hugepages-2Mi"):    resource.MustParse("1Gi"),
 							},
 							Limits: api.ResourceList{
-								api.ResourceName("hugepages-2Mi"): resource.MustParse("1Gi"),
+								api.ResourceName(api.ResourceCPU):    resource.MustParse("10"),
+								api.ResourceName(api.ResourceMemory): resource.MustParse("10G"),
+								api.ResourceName("hugepages-2Mi"):    resource.MustParse("1Gi"),
```
Comment (tianshapjq) [inline pkg/api/validation/validation_test.go]: would you please add some tests about ignoring the cpu or memory and raise the errors?

File: pkg/api/validation/validation.go (hunk 1 above)
Comment (tanshanshan) [reply pkg/api/validation/validation.go]: if we do not specify the cpu, memory and hugepages , reqContainsHugePages and limContainsHugePages are both false , so the if condition is false .

**Comment by lmdaly** [comment]
Wondering on the use case on why a container would request CPU and huge pages but not memory? Or what the requirement for a container needing to request CPU with huge pages is?

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
i would prefer we do this as follows:

remove hugepages from list of supported qos resources.

https://github.com/kubernetes/kubernetes/blob/master/pkg/api/helper/qos/qos.go#L28

https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/helper/qos/qos.go#L29

during validation, return an error if any pod uses hugepages, but reports to be BestEffort.

https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/helper/qos/qos.go#L38

**Comment by tanshanshan** [comment]
the first two points
remove hugepages from list of supported qos resources.
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/helper/qos/qos.go#L28
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/v1/helper/qos/qos.go#L29

the last one 
`during validation, return an error if any pod uses hugepages, but reports to be BestEffort.`

Could you tell me where to validate , not in `pkg/api/validation/validation.go` .

Thanks for your help,  @derekwaynecarr .

**Comment by derekwaynecarr** [comment]
@tanshanshan if you can update this pr to make the changes to qos files, and leave the validation check as is (just verify that a request was made for cpu or memory) then i think this is fine.

@lmdaly the intent is that workloads that desire hugepages as best practice should not be in the BestEffort tier.  From the kubelet code perspective, I would like to keep the pod level cgroup unbounded for all workloads in that part of the tree.

**Review by derekwaynecarr — APPROVED** [review]

**Comment by derekwaynecarr** [comment]
/kind bug

similar to gpus, we decided hugepages should not impact qos.  having hugepages impact qos would have made it existing G pods would no longer be G (as they lacked a request for hugepages).  hugepages also vary in size, so enumerating a request for each potential size makes no sense either.  as a result, hugepages are similar to gpus or other devices and should be outside the qos tree.

File: pkg/apis/core/helper/qos/qos.go [hunk 3]
```diff
@@ -22,12 +22,11 @@ import (
 	"k8s.io/apimachinery/pkg/api/resource"
 	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/kubernetes/pkg/apis/core"
-	"k8s.io/kubernetes/pkg/apis/core/helper"
 )
 
 func isSupportedQoSComputeResource(name core.ResourceName) bool {
 	supportedQoSComputeResources := sets.NewString(string(core.ResourceCPU), string(core.ResourceMemory))
```
Comment (thockin) [inline pkg/apis/core/helper/qos/qos.go]: this should be a var outside the functin

Comment (thockin) [reply pkg/apis/core/helper/qos/qos.go]: same in other places

**Comment by tanshanshan** [comment]
@@thockin @smarterclayton  @derekwaynecarr  PTAL ,thanks

**Comment by derekwaynecarr** [comment]
we can queue this up for the 1.10 stream, and ensure it lands before merging https://github.com/kubernetes/kubernetes/pull/56939

**Pull request merged**